buildSrc/subprojects/performance/src/main/groovy/org/gradle/testing/DistributedPerformanceTest.groovy

```diff
@@ -338,35 +339,35 @@ class DistributedPerformanceTest extends PerformanceTest {
                     it.@name.text() == zipName
                 }
                 if (fileNode) {
-                    def resultsDirectory = new File(workerTestResultsTempDir, jobId)
                     def contentUri = fileNode.content.@href.text()
                     client.get(path: contentUri, contentType: ContentType.BINARY) {
                         resp, inputStream ->
-                            unzippedFiles = unzipToDirectory(inputStream, resultsDirectory)
+                            unzippedJUnitXmls = parseXmlsInZip(inputStream)
                     }
                 }
             }
         }
-        unzippedFiles
+        unzippedJUnitXmls
     }
 
-    @TypeChecked(TypeCheckingMode.SKIP)
-    def unzipToDirectory(inputStream, destination) {
-        def unzippedFiles = []
+    List<JUnitTestSuite> parseXmlsInZip(InputStream inputStream) {
+        List<JUnitTestSuite> parsedXmls = []
         new ZipInputStream(inputStream).withStream { zipInput ->
             def entry
             while (entry = zipInput.nextEntry) {
-                if (!entry.isDirectory()) {
-                    def file = new File(destination, entry.name)
-                    file.parentFile?.mkdirs()
-                    new FileOutputStream(file).withStream {
-                        it << zipInput
+                if (!entry.isDirectory() && entry.name.endsWith('.xml')) {
+                    PipedOutputStream os = new PipedOutputStream()
+                    PipedInputStream is = new PipedInputStream(os)
+                    Thread.start {
```

Comment:
Why do we start a new thread? That doesn't seem like a good idea...

The zip file input stream is also not thread safe I think.

Reply:
Why do you need to do this? Just replace this whole thing with:
```
                if (!entry.isDirectory() && entry.name.endsWith('.xml')) {
                    parsedXmls.add(JUnitMarshalling.unmarshalTestSuite(zipInput))
                }
```

Does this work?

Reply:
No it doesn't. Actually, this is also what I tried at first, but `unmarshalTestSuite` will close `zipInput` after it finishes, and in next while loop `zipInput` will throw `InputAlreadyClosedException`. We can sure use `File` as a intermediate, but I think that's unnecessary.

Reply:
Then you need to wrap `zipInput` in a `CloseShieldInputStream`.

Reply:
Great to know this! Thanks!